src/errors.rs

```diff
@@ -0,0 +1,123 @@
+use crate::{
+    protos::coresdk::activity_result::ActivityResult,
+    protos::coresdk::workflow_completion::WfActivationCompletion,
+    protos::temporal::api::workflowservice::v1::PollWorkflowTaskQueueResponse,
+    workflow::WorkflowError,
+};
+use tonic::codegen::http::uri::InvalidUri;
+
+pub(crate) struct ShutdownErr;
+pub(crate) struct WorkflowUpdateError {
+    /// Underlying workflow error
+    pub source: WorkflowError,
+    /// The run id of the erring workflow
+    pub run_id: String,
+}
+
+/// Errors thrown during initialization of [crate::Core]
+#[derive(thiserror::Error, Debug, displaydoc::Display)]
+pub enum CoreInitError {
+    /// Invalid URI: {0:?}
+    InvalidUri(#[from] InvalidUri),
+    /// Server connection error: {0:?}
+    TonicTransportError(#[from] tonic::transport::Error),
+}
+
+/// Errors thrown by [crate::Core::poll_workflow_task]
+#[derive(thiserror::Error, Debug, displaydoc::Display)]
+pub enum PollWfError {
+    /// There was an error specific to a workflow instance with id ({run_id}): {source:?}
+    WorkflowUpdateError {
```

Comment:
I'm assuming this means kill the workflow cache for given `run_id`? If so, can we move this to an activation (e.g. `DeleteWorkflow{ reason }`.

Reply:
I don't like it so much. This call already needs to return a `Result`, so it's not like we really gain anything by moving the error into the activation, and we can't make activations with `DeleteWorkflow` in them be the only way of forcing that action because problems can happen when you call the completion API too. So, the path of "delete the workflow because of an error from an API call" always needs to exist, and that feedback is immediate and obvious vs the somewhat indirect way of stuffing it in an activation.

Reply:
You can always translate this error into something you stuff inside an activation if you want to do that for your control flow purposes